File: Python/Minimum Size Subarray Sum.py
from sys import maxsize
from typing import List


class Solution:
    """
    Time:   O(n)
    Memory: O(1)
    """

    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        min_len = maxsize
        window_sum = start = 0

        for end, num in enumerate(nums):
            window_sum += num
            while window_sum >= target:
                min_len = min(min_len, end - start + 1)
                window_sum -= nums[start]
                start += 1

        return min_len if min_len != maxsize else 0

File: Java/Minimum Size Subarray Sum.java
class Solution {

    /**
     * Time:   O(n)
     * Memory: O(1)
     */
    public int minSubArrayLen(int target, int[] nums) {
        int minLen = Integer.MAX_VALUE;
        int window_sum = 0, start = 0;

        for (int end = 0; end < nums.length; ++end) {
            window_sum += nums[end];

            while (window_sum >= target) {
                minLen = Math.min(minLen, end - start + 1);
                window_sum -= nums[start++];
            }
        }

        return minLen == Integer.MAX_VALUE ? 0 : minLen;
    }
}
